design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (i.e. by following syntax 
+like Hive's `NAMED_STRCUT` construct). 
+
+Future extensions: 
+* Support some`AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+
+#### DELETE FROM
+
+```sql 
+DELETE FROM table_name WHERE condition;
+```
+
+The `DELETE` statement can be used to insert tombstones into kafka topics that back tables. If the
+source is not back a table, the statement will fail. For this KLIP, we will only support conditions
+of the form `ROWKEY = value`. While this may initially be confusing (as SQL standard allows deleting
+on any arbitrary condition), we have precedent for this in stream-table joins (the join criteria
+may only be on the key of the table).
+
+#### CREATE STREAM/TABLE
+
+```sql
+CREATE STREAM stream_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+```sql
+CREATE TABLE table_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+The `WITH` clause will now accept two additional property_names: `PARTITIONS`/`REPLICAS` If the 
+topic *does not exist* and these values are specified, this command will create a topic with a the
+set number of partitions and replicas before executing the command. If the topic exists and has
+a different number of partitions/replicas, this command will fail.
+
+This functionality will be flagged under a configuration `ksql.ddl.source.topic.create.allow`,
+which will default to `true` (new behavior is the default).
+
+## Design
+
+### API Design
+
+The above APIs make various trade-offs, outlined below:
+
+* The proposal to overload `INSERT INTO` is to remain as compatible with SQL standard as possible.
+The parser should be lookahead and determine whether or not `VALUES` is present in order to leverage
+that data producing APIs. An alternative would be to use the word `PRODUCE`, which also makes it
+clear that it uses low-level Kafka producers.
+* The proposal to use `DELETE` encompasses various trade-offs:
+  * We need a different mechanism to produce tombstones than `INSERT INTO` because there is no good
+  way with the standard SQL to differentiate between an entirely `NULL` value (i.e. tombstone) and 
+  `NULL` values for every column.
+  * Introducing a new syntax (e.g. `INSERT TOMBSTONE`) will require users to understand another
+  language construct and make KSQL depart further from SQL standard
+  * This is a step closer to SQL standard, and we can always support more `DELETE` conditions in
+  the future
+* Allowing `CREATE TABLE/STREAM` to specify a kafka topic that does not yet exist makes it easy to
```

Comment:
How about a new property `CREATE_TOPIC`, to complement the existing `KAFKA_TOPIC` property in the `WITH` clause? This makes it explicit that you want to create this topic if it doesn't exist, and will require the presence of `PARTITIONS` and `REPLICAS`.

Alternately, We could introduce a new type of compound property like `CREATE_WITH={ partitions: <partitions>, replicas=<replicas>}` so that the partitions and replicas can only be set when you are creating topics, and also _must_ be set when you are creating topics. In this case, we would also retain `KAFKA_TOPIC` as a property, with `CREATE_WITH` being an additional clause to specify creating a topic with certain properties. 

Something like this seems more organized to me.

Reply:
I think I agree with @rmoff that introducing new syntax that differs between the CT/CS and CTAS/CSAS will lead to more confusion. 

I do believe, though, that we should make it explicit for CT/CS statements that you intend to create an empty topic - this will help catch typos if you intended to create a stream/table off of an existing topic (we don't want that type of behavior enabled for "production" use cases and definitely not for headless mode). I think a global config is sufficient here (instead of having it configured per statement) because it seems that you would either want all CT/CS statements to optionally create topics (testing), or none of them (production).

Reply:
So the proposal is to have a global config that allows CT/CS statements to create topics. And if both `PARTITONS` and `REPLICAS` are not specified, then the topic creation attempt will be failed with a sensible error message?